src/iamserver/protectedmessagehandler.cpp

```diff
@@ -0,0 +1,484 @@
+}
+
+template <typename Message>
+static void SetErrorInfo(Message& message, const aos::Error& error)
+{
+    *message.mutable_error() = ConvertAosErrorToProto(error);
+}
+
+static aos::InstanceIdent ConvertToAOS(const iamproto::InstanceIdent& val)
+{
+    aos::InstanceIdent result;
+
+    result.mServiceID = val.service_id().c_str();
+    result.mSubjectID = val.subject_id().c_str();
+    result.mInstance  = val.instance();
+
+    return result;
+}
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+aos::Error ProtectedMessageHandler::Init(NodeController& nodeController,
+    aos::iam::identhandler::IdentHandlerItf& identHandler, aos::iam::permhandler::PermHandlerItf& permHandler,
+    aos::iam::NodeInfoProviderItf& nodeInfoProvider, aos::iam::nodemanager::NodeManagerItf& nodeManager,
+    aos::iam::provisionmanager::ProvisionManagerItf& provisionManager)
+{
+    return PublicMessageHandler::Init(
+        nodeController, identHandler, permHandler, nodeInfoProvider, nodeManager, provisionManager);
+}
+
+void ProtectedMessageHandler::RegisterServices(grpc::ServerBuilder& builder, bool provisionMode)
+{
+    PublicMessageHandler::RegisterServices(builder);
+
+    if (GetPermHandler() != nullptr) {
+        builder.RegisterService(static_cast<iamproto::IAMPermissionsService::Service*>(this));
+    }
+
+    if (IsMainNode()) {
+        builder.RegisterService(static_cast<iamproto::IAMCertificateService::Service*>(this));
+
+        if (provisionMode) {
+            builder.RegisterService(static_cast<iamproto::IAMProvisioningService::Service*>(this));
+        }
+
+        builder.RegisterService(static_cast<iamproto::IAMNodesService::Service*>(this));
+    }
+}
+
+void ProtectedMessageHandler::Close()
+{
+    PublicMessageHandler::Close();
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+/***********************************************************************************************************************
+ * IAMPublicNodesService implementation
+ **********************************************************************************************************************/
+
+grpc::Status ProtectedMessageHandler::RegisterNode(grpc::ServerContext*                         context,
+    grpc::ServerReaderWriter<::iamproto::IAMIncomingMessages, ::iamproto::IAMOutgoingMessages>* stream)
+{
+    LOG_DBG() << "Register node has been received on the protected channel";
+
+    using aos::NodeStatus;
+    using aos::NodeStatusEnum;
+
+    static const std::vector<NodeStatus> allowedStatuses = {NodeStatusEnum::eProvisioned, NodeStatusEnum::ePaused};
```

Comment:
for my opinion it could be better to move it in const section, and in theory it could be std::array instead of std::vector